include/nbl/builtin/hlsl/cpp_compat/impl/intrinsics_impl.hlsl

```diff
@@ -3,428 +3,573 @@
+		return spirv::SPIRV_FUNCTION_NAME<T>(arg);\
 	}\
-};\
-
-DEFINE_FIND_MSB_COMMON_SPECIALIZATION(int16_t, int32_t)
-DEFINE_FIND_MSB_COMMON_SPECIALIZATION(uint16_t, uint32_t)
-#ifndef __HLSL_VERSION
-DEFINE_FIND_MSB_COMMON_SPECIALIZATION(int8_t, int32_t)
-DEFINE_FIND_MSB_COMMON_SPECIALIZATION(uint8_t, uint32_t)
-#endif
+};
 
-template<>
-struct find_msb_helper<uint64_t>
+#define FIND_MSB_LSB_RETURN_TYPE conditional_t<is_vector_v<T>, vector<int32_t, vector_traits<T>::Dimension>, int32_t>;
+AUTO_SPECIALIZE_TRIVIAL_CASE_HELPER(find_msb_helper, findUMsb, FIND_MSB_LSB_RETURN_TYPE)
+AUTO_SPECIALIZE_TRIVIAL_CASE_HELPER(find_msb_helper, findSMsb, FIND_MSB_LSB_RETURN_TYPE)
+AUTO_SPECIALIZE_TRIVIAL_CASE_HELPER(find_lsb_helper, findILsb, FIND_MSB_LSB_RETURN_TYPE)
+#undef FIND_MSB_LSB_RETURN_TYPE
+AUTO_SPECIALIZE_TRIVIAL_CASE_HELPER(bitReverse_helper, bitReverse, T)
+AUTO_SPECIALIZE_TRIVIAL_CASE_HELPER(transpose_helper, transpose, T)
+AUTO_SPECIALIZE_TRIVIAL_CASE_HELPER(length_helper, length, T)
+AUTO_SPECIALIZE_TRIVIAL_CASE_HELPER(normalize_helper, normalize, T)
+AUTO_SPECIALIZE_TRIVIAL_CASE_HELPER(rsqrt_helper, inverseSqrt, T)
+
+template<typename UInt64> NBL_PARTIAL_REQ_TOP(is_same_v<UInt64, uint64_t>)
+struct find_msb_helper<UInt64 NBL_PARTIAL_REQ_BOT(is_same_v<UInt64, uint64_t>) >
 {
-	static int32_t findMSB(NBL_CONST_REF_ARG(uint64_t) val)
+	using return_t = int32_t;
+	static return_t __call(NBL_CONST_REF_ARG(UInt64) val)
 	{
-#ifdef __HLSL_VERSION
 		const uint32_t highBits = uint32_t(val >> 32);
-		const int32_t highMsb = find_msb_helper<uint32_t>::findMSB(highBits);
+		const int32_t highMsb = find_msb_helper<uint32_t>::__call(highBits);
 
 		if (highMsb == -1)
 		{
 			const uint32_t lowBits = uint32_t(val);
-			const int32_t lowMsb = find_msb_helper<uint32_t>::findMSB(lowBits);
+			const int32_t lowMsb = find_msb_helper<uint32_t>::__call(lowBits);
 			if (lowMsb == -1)
 				return -1;
 
 			return lowMsb;
 		}
 
 		return highMsb + 32;
-#else
-		return glm::findMSB(val);
-#endif
 	}
 };
+template<typename UInt64> NBL_PARTIAL_REQ_TOP(is_same_v<UInt64, uint64_t>)
+struct find_lsb_helper<UInt64 NBL_PARTIAL_REQ_BOT(is_same_v<UInt64, uint64_t>) >
+{
+	static int32_t __call(NBL_CONST_REF_ARG(uint64_t) val)
+	{
+		const uint32_t lowBits = uint32_t(val);
+		const int32_t lowLsb = find_lsb_helper<uint32_t>::__call(lowBits);
+
+		if (lowLsb == -1)
+		{
+			const uint32_t highBits = uint32_t(val >> 32);
+			const int32_t highLsb = find_lsb_helper<uint32_t>::__call(highBits);
+			if (highLsb == -1)
+				return -1;
+			else
+				return 32 + highLsb;
+		}
+
+		return lowLsb;
+	}
+};
+
+#define AUTO_SPECIALIZE_TRIVIAL_CASE_HELPER_2_ARG_FUNC(HELPER_NAME, SPIRV_FUNCTION_NAME, RETURN_TYPE)\
+template<typename T> NBL_PARTIAL_REQ_TOP(always_true<decltype(spirv::SPIRV_FUNCTION_NAME<T>(experimental::declval<T>()))>)\
+struct HELPER_NAME<T NBL_PARTIAL_REQ_BOT(always_true<decltype(spirv::SPIRV_FUNCTION_NAME<T>(experimental::declval<T>()))>) >\
```

Comment:
will never match and will silently fail, you're calling with only one argument worth of declval

btw have you thought about
```cpp
#define DECLVAL(z,n,_T) experimental::declval<_T>()
#define DECL_ARG(z,n,_T) const _T arg##n
#define ARG(z,n,_T) arg##n
#define AUTO_SPECIALIZE_TRIVIAL_CASE_HELPER(HELPER_NAME,SPIRV_FUNCTION_NAME,RETURN_TYPE,ARG_C)
```

then use `BOOST_PP_ENUM(ARG_C,DECLVAL,T)` for the declvals and likewise for the argument definition and usage?

Reply:
fixed

Reply:
not implementing it right now